File: samples/decode_mot_encode_multi/decode_mot_encode_multi.cpp
/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <filesystem>
#include <fstream>
#include <iostream>

#include "bytetracker.hpp"
#include "common/cmdline.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>(
      "model_prefix", 'm', "model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/"
      "bytetrack_m_mot17-int8-percentile-1_3_800_1440-vacc-pipeline/mod");
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "./data/configs/yolo_div255_yuv_nv12.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<float>("detect_thresh", '\0', "threshold for detection", false, 0.5);

  args.add<float>("track_thresh", '\0', "threshold for tracker", false, 0.6);
  args.add<uint32_t>("track_buffer", '\0', "the frames for keep lost tracks",
                     false, 30);
  args.add<float>("match_thresh", '\0', "matching threshold for tracking",
                  false, 0.9);
  args.add<float>("min_box_area", '\0', "filter out tiny boxes", false, 100);

  args.add<std::string>("uri", '\0', "input uri", false,
                        "../data/videos/test.mp4");
  args.add<std::string>("output_path", '\0', "output_path", false, "");
  args.add<uint32_t>("num_channels", '\0',
                     "number of channels to decode [ default: 1 ]", false, 1);
  args.add<uint32_t>("loop", '\0', "loop count for each channel [ default: 1 ]",
                     false, 1);
  args.add<std::string>("uri_list", '\0', "input uri list file", false, "");
  args.add<uint32_t>("disable_encoder", '\0', "enable encoder or not", false,
                     0);
  args.parse_check(argc, argv);
  return args;
}

int Process(const cmdline::parser& args, std::string input_uri,
            uint32_t cur_device_id, uint32_t index = 0, uint32_t loop = 1) {
  // initialize device
  CHECK(vsx::SetDevice(cur_device_id) == 0)
      << "Failed to set device id: " << cur_device_id;

  auto detect_thresh = args.get<float>("detect_thresh");
  auto track_buffer = args.get<uint32_t>("track_buffer");
  auto track_thresh = args.get<float>("track_thresh");

  for (uint32_t inn = 0; inn < loop; inn++) {
    std::cout << "Process " << index << " on device " << cur_device_id
              << " for uri:" << input_uri << ", loop: " << inn << std::endl;
    // initialize model
    int batch_size = 1;
    int fps = 30;
    vsx::ByteTracker tracker(args.get<std::string>("model_prefix"),
                             args.get<std::string>("vdsp_params"), batch_size,
                             cur_device_id, detect_thresh, track_buffer,
                             track_thresh, fps);
    // open uri
    vsx::VideoCapture cap(input_uri, vsx::FULLSPEED_MODE, cur_device_id, true,
                          false);
    CHECK(cap.isOpened()) << "Failed to open uri: " << input_uri;

    vsx::Image frame;  // frame format is nv12, memory is in device
    std::shared_ptr<vsx::FrameAttr> frame_attr;
    if (!cap.read(frame, frame_attr)) {  // get first frame
      std::cout << "Failed to read frame\n";
      return -1;
    }

    // print video info
    // std::cout << "Frame width: " << frame.Width()
    //           << ", width_pitch: " << frame.WidthPitch()
    //           << ", height: " << frame.Height()
    //           << ", height_pitch: " << frame.HeightPitch() << std::endl;
    // std::cout << "Frame rate: " << frame_attr->video_fps << std::endl;
    // std::cout << "Video codec_info: " << frame_attr->codec_info << std::endl;
    // if (frame_attr->color_space == vsx::ImageColorSpace::kCOLOR_SPACE_BT709)
    //   std::cout << "Frame ColorSpace:  BT709 \n";
    // else if (frame_attr->color_space ==
    //          vsx::ImageColorSpace::kCOLOR_SPACE_BT601)
    //   std::cout << "Frame ColorSpace:  BT601 \n";
    // else if (frame_attr->color_space ==
    //          vsx::ImageColorSpace::kCOLOR_SPACE_BT709_LIMIT_RANGE)
    //   std::cout << "Frame ColorSpace:  BT709_LIMIT_RANGE \n";
    // else if (frame_attr->color_space ==
    //          vsx::ImageColorSpace::kCOLOR_SPACE_BT601_FULL_RANGE)
    //   std::cout << "Frame ColorSpace:  BT601_FULL_RANGE \n";

    auto disable_encoder = args.get<uint32_t>("disable_encoder");
    std::unique_ptr<vsx::VideoWriter> video_writer;

    // create video_writer
    if (!disable_encoder) {
      float frame_rate = frame_attr->video_fps;

      vsx::CodecType codec_type;
      if (frame_attr->codec_info.find("avc1") != std::string::npos) {
        codec_type = vsx::CODEC_TYPE_H264;
      } else if (frame_attr->codec_info.find("hevc") != std::string::npos) {
        codec_type = vsx::CODEC_TYPE_HEVC;
      } else {
        std::cerr << "undefined codec_type:" << frame_attr->codec_info
                  << std::endl;
        return -1;
      }

      // create video writer
      auto output_path = args.get<std::string>("output_path");
      auto outfile_name =
          output_path + "/channel_" + std::to_string(index) + ".ts";
      video_writer = std::make_unique<vsx::VideoWriter>(
          outfile_name, frame_rate, codec_type, 0, 0, cur_device_id);
    }

    // detection result file
    auto output_path = args.get<std::string>("output_path");
    auto track_file_name =
        output_path + "/channel_" + std::to_string(index) + ".txt";
    std::ofstream track_file(track_file_name);
    if (!track_file.is_open()) {
      std::cout << "Error, Failed to open track result file: "
                << track_file_name << std::endl;
      return -1;
    }

    uint32_t frame_id = 0;
    do {
      frame_id++;
      // inference
      auto result = tracker.Process(frame);
      auto res_shape = result.Shape();
      const float* res_data = result.Data<float>();
      std::vector<vsx::Rect> obj_rects;
      for (int i = 0; i < res_shape[0]; i++) {
        if (res_data[0] < 0) break;

        vsx::Rect rect;
        rect.x = int(res_data[2]);
        rect.y = int(res_data[3]);
        rect.w = int(res_data[4]);
        rect.h = int(res_data[5]);
        obj_rects.push_back(rect);

        float score = res_data[1];
        // '{frame},{id},{x1},{y1},{w},{h},{s},-1,-1,-1\n'
        track_file << frame_id << "," << static_cast<int>(res_data[6]) << ","
                   << std::setiosflags(std::ios::fixed) << std::setprecision(2)
                   << res_data[2] << "," << res_data[3] << "," << res_data[4]
                   << "," << res_data[5] << "," << score << ",-1,-1,-1\n";
        res_data += vsx::kDetectionOffset;
      }
      if (!disable_encoder) {
        // draw box to image, ONLY support YUV_NV12 format image
        vsx::Image output_image;
        vsx::DrawRectangle(frame, output_image, obj_rects,
                           vsx::YUVColor::YUV_RED, 2);
        video_writer->write(output_image, frame_attr);
      }

    } while (cap.read(frame, frame_attr));

    std::cout << "Close cap\n";
    cap.release();
    track_file.close();

    if (!disable_encoder) {
      std::cout << "Close video writer\n";
      video_writer->release();
    }
  }
  return 0;
}

int main(int argc, char* argv[]) {
  // get command parameters
  auto args = ArgumentParser(argc, argv);
  auto num_channels = args.get<uint32_t>("num_channels");
  // get input uri
  std::vector<std::string> input_list;
  if (!args.get<std::string>("uri_list").empty())
    // load uri list from file
    input_list = vsx::LoadLabels(args.get<std::string>("uri_list"));
  else if (!args.get<std::string>("uri").empty()) {
    for (uint32_t i = 0; i < num_channels; i++) {
      input_list.push_back(args.get<std::string>("uri"));
    }
  }

  if (args.get<std::string>("output_path").empty()) {
    std::cerr << "please set output_path \n";
    return -1;
  }
  std::filesystem::path output_path(args.get<std::string>("output_path"));
  std::error_code ec;
  std::filesystem::create_directories(output_path, ec);
  if (ec) {
    std::cerr << "Failed to create output_path "
              << args.get<std::string>("output_path")
              << ", message:" << ec.message() << std::endl;
    return -1;
  }

  assert(input_list.size() == num_channels);

  auto device_id = args.get<uint32_t>("device_id");
  auto loop = args.get<uint32_t>("loop");

  uint32_t split_num = num_channels / 2;
  if (split_num == 0) split_num = 1;

  std::vector<std::thread> process_list;
  process_list.reserve(num_channels);
  for (size_t i = 0; i < num_channels; i++) {
    auto cur_device_id = device_id + int(i / split_num);
    std::cout << "input_uri:" << input_list[i]
              << ", device_id: " << cur_device_id << std::endl;
    process_list.emplace_back(&Process, std::cref(args), input_list[i],
                              cur_device_id, i, loop);
  }

  for (auto& p : process_list) {
    p.join();
  }
}

File: samples/decode_mot_encode_multi/decode_mot_encode_multi.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.detector import Detector
import vaststreamx as vsx
import multiprocessing
from tracker.byte_tracker import BYTETracker
import argparse
import shutil


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m", "--model_prefix", default="", help="model prefix of the model suite files"
    )
    parser.add_argument(
        "--vdsp_params", default="", help="vdsp preprocess parameter file"
    )
    parser.add_argument(
        "-d", "--device_id", default=0, type=int, help="device id to run"
    )
    parser.add_argument(
        "--detect_thresh", default=0.01, type=float, help="detector threshold"
    )
    # tracker params
    parser.add_argument(
        "--track_thresh", type=float, default=0.6, help="tracking confidence threshold"
    )
    parser.add_argument(
        "--track_buffer", type=int, default=30, help="the frames for keep lost tracks"
    )
    parser.add_argument(
        "--match_thresh",
        type=float,
        default=0.9,
        help="matching threshold for tracking",
    )
    parser.add_argument(
        "--min_box_area", type=float, default=100, help="filter out tiny boxes"
    )
    parser.add_argument(
        "--uri",
        type=str,
        required=True,
        help="uri to decode",
    )
    parser.add_argument(
        "--output_path",
        default="output_result",
        type=str,
        help="output path [ default: output_result ]",
    )
    parser.add_argument(
        "--num_channels",
        default=1,
        type=int,
        help="number of channels to decode [ default: 1 ]",
    )
    parser.add_argument(
        "--loop",
        default=1,
        type=int,
        help="loop count for each channel [ default: 1 ]",
    )
    parser.add_argument("--uri_list", default="", help="input uri list file")
    parser.add_argument(
        "--disable_encoder", action="store_true", help="enable encoder or not"
    )
    parser.add_argument("--save_output", action="store_true", help="save output or not")

    args = parser.parse_args()
    return args

def process(args, input_uri, cus_device_id, index=0, loop=1):
    vsx.set_device(cus_device_id)
    for inn in range(loop):
        print(
            f"Process {index} on device {cus_device_id} for uri:{input_uri}, loop {inn}"
        )
        batch_size = 1
        detector = Detector(
            args.model_prefix, args.vdsp_params, batch_size, cus_device_id
        )
        detector.set_threshold(args.detect_thresh)
        tracker = BYTETracker(
            args.track_thresh + 0.1, args.track_thresh, args.track_buffer, args.match_thresh
        )

        cap = vsx.VideoCapture(
            input_uri, vsx.CaptureMode.FULLSPEED_MODE, cus_device_id, True
        )

        ret, frame, frame_attr = cap.read()
        assert ret

        frame_rate = int(frame_attr.video_fps)
        if frame_attr.codec_info.find("avc1") != -1:
            codec_type = vsx.CODEC_TYPE_H264
        elif frame_attr.codec_info.find("hevc") != -1:
            codec_type = vsx.CODEC_TYPE_HEVC
        else:
            print(f"undefined codec_type:{frame_attr.codec_info}")
            exit(-1)

        writer = None
        if not args.disable_encoder:
            default_bit_rate = 8000000
            key_frame_interval = frame_rate
            writer = vsx.VideoWriter(
                args.output_path + f"/channel_{index}.ts",
                frame_rate,
                codec_type,
                default_bit_rate,
                key_frame_interval,
                cus_device_id,
            )
        
        text_file = open(args.output_path + f"/channel_{index}.txt","wt")
        save_format = "{frame},{id},{x1},{y1},{w},{h},{s},-1,-1,-1\n"

        frame_id = 0
        while True:
            frame_id += 1
            objects = detector.process(frame)
            track_targets = tracker.update(objects)
            track_tlwhs = []
            track_ids = []
            track_scores = []
            bboxes=[]
            for t in track_targets:
                tlwh = t.tlwh
                tid = t.track_id
                vertical = tlwh[2] / tlwh[3] > 1.6
                if tlwh[2] * tlwh[3] > args.min_box_area and not vertical:
                    track_tlwhs.append(tlwh)
                    track_ids.append(tid)
                    track_scores.append(float(t.score))
                    bboxes.append([int(tlwh[0]),int(tlwh[1]),int(tlwh[2]),int(tlwh[3])])
            # write video frame
            if writer:
                vsx_image = vsx.draw_rectangle(frame, bboxes, vsx.YUV_RED, 2, True)
                writer.write(vsx_image, frame_attr)
            # write track result
            for tlwh, track_id, score in zip(track_tlwhs, track_ids, track_scores):
                if track_id < 0:
                    continue
                x1, y1, w, h = tlwh
                line = save_format.format(
                    frame=frame_id,
                    id=track_id,
                    x1=round(x1, 1),
                    y1=round(y1, 1),
                    w=round(w, 1),
                    h=round(h, 1),
                    s=round(score, 2),
                )
                text_file.write(line)

            ret, frame, frame_attr = cap.read()
            if ret == 0:
                break

        cap.release()
        if writer:
            writer.release()
        text_file.close()

if __name__ == "__main__":
    args = argument_parser()
    process_list = []

    input_list = []
    if args.uri_list != "":
        with open(args.uri_list, "rt") as f:
            input_list = [line.strip() for line in f.readlines()]
        print(f"input_list:{input_list}")
    elif args.uri != "":
        for i in range(args.num_channels):
            input_list.append(args.uri)
    if args.save_output:
        if args.output_path == "":
            print("please set output_path")
            exit(-1)
        if os.path.exists(args.output_path):
            shutil.rmtree(args.output_path)
        os.makedirs(args.output_path, exist_ok=True)
    assert len(input_list) == args.num_channels

    device_id = args.device_id
    split_num = args.num_channels / 2
    if split_num == 0:
        split_num = 1

    for index, input_uri in enumerate(input_list):
        cus_device_id = device_id + int(index / split_num)
        print(f"input_uri:{input_uri},device_id:{cus_device_id}")
        p = multiprocessing.Process(
            target=process, args=(args, input_uri, cus_device_id, index, args.loop)
        )
        process_list.append(p)
        p.start()

    for p in process_list:
        p.join()